service/src/main/java/com/epam/esm/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,53 @@
+package com.epam.esm.impl;
+
+import com.epam.esm.TagDao;
+import com.epam.esm.TagService;
+import com.epam.esm.exception.EntityNotFoundException;
+import com.epam.esm.exception.IncorrectEntityException;
+import com.epam.esm.validator.TagValidator;
+import com.epam.esm.validator.ValidationError;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class TagServiceImpl implements TagService {
+    private TagDao tagDao;
+    private TagValidator validator;
+
+    @Autowired
+    public TagServiceImpl(TagDao tagDao, TagValidator validator) {
+        this.tagDao = tagDao;
+        this.validator = validator;
+    }
+
+    @Override
+    public List<Tag> findAll() {
+        return tagDao.findAll();
+    }
+
+    @Override
+    public Optional<Tag> findById(Long id) {
+        Optional<Tag> optionalTag = tagDao.findById(id);
+        if (!optionalTag.isPresent()) {
+            throw new EntityNotFoundException("Tag had not found by id.", id);
+        }
+        return tagDao.findById(id);
+    }
+
+    @Override
+    public Tag save(Tag tag) {
+        List<ValidationError> validationErrors = validator.validateTagName(tag.getName());
+        if (!validationErrors.isEmpty()) {
+            throw new IncorrectEntityException("You have problem with input parameters.", validationErrors);
```

Comment:
> You have problem with...
better to avoid direct messages, better do describe the problem e.g. 'input parameters are wrong ....'

Reply:
Fixed